### ISSUE TITLE
--help broken for subcommands by #1804

### ISSUE THREAD

**Issue by cocreature** [commercialhaskell/stack#1887]
#1804 seems to have broken `--help` for commands. Previously (17a62d01c40d9c3ff55be9f7c98bffb10973f2a4) `stack build --help` resulted in the following output
```
Usage: stack build [--trace] [--profile] [TARGET] [--[no-]library-profiling]
                   [--[no-]executable-profiling] [--[no-]haddock]
                   [--[no-]haddock-deps] [--dry-run] [--pedantic] [--fast]
                   [--ghc-options OPTION] [--flag PACKAGE:[-]FLAG]
                   [--[no-]copy-bins] [--prefetch] ([--dependencies-only] |
                   [--only-snapshot] | [--only-dependencies]) ([--file-watch] |
                   [--file-watch-poll]) [--[no-]keep-going] [--force-dirty]
                   [--[no-]test] [--[no-]rerun-tests]
                   [--test-arguments TEST_ARGS] [--coverage] [--no-run-tests]
                   [--[no-]bench] [--benchmark-arguments BENCH_ARGS]
                   [--no-run-benchmarks] [--exec CMD [ARGS]] [--only-configure]
                   [--reconfigure] [--cabal-verbose] [--help]
  Build the package(s) in this directory/configuration

Available options:
  --trace                  Enable profiling in libraries, executables, etc. for
                           all expressions and generate a backtrace on exception
  --profile                Enable profiling in libraries, executables, etc. for
                           all expressions and generate a profiling report in
                           exec or benchmarks
  TARGET                   If none specified, use all packages
  --[no-]library-profiling Enable/disable library profiling for TARGETs and all
                           its dependencies
  --[no-]executable-profiling
                           Enable/disable executable profiling for TARGETs and
                           all its dependencies
  --[no-]haddock           Enable/disable generating Haddocks the package(s) in
                           this directory/configuration
  --[no-]haddock-deps      Enable/disable building Haddocks for dependencies
  --dry-run                Don't build anything, just prepare to
  --pedantic               Turn on -Wall and -Werror
  --fast                   Turn off optimizations (-O0)
  --ghc-options OPTION     Additional options passed to GHC
  --flag PACKAGE:[-]FLAG   Override flags set in stack.yaml (applies to local
                           packages and extra-deps)
  --[no-]copy-bins         Enable/disable copying binaries to the local-bin-path
                           (see 'stack path')
  --prefetch               Fetch packages necessary for the build immediately,
                           useful with --dry-run
  --dependencies-only      A synonym for --only-dependencies
  --only-snapshot          Only build packages for the snapshot database, not
                           the local database
  --only-dependencies      Only build packages that are dependencies of targets
                           on the command line
  --file-watch             Watch for changes in local files and automatically
                           rebuild. Ignores files in VCS boring/ignore file
  --file-watch-poll        Like --file-watch, but polling the filesystem instead
                           of using events
  --[no-]keep-going        Enable/disable continue running after a step fails
                           (default: false for build, true for test/bench)
  --force-dirty            Force treating all local packages as having dirty
                           files (useful for cases where stack can't detect a
                           file change)
  --[no-]test              Enable/disable testing the package(s) in this
                           directory/configuration
  --[no-]rerun-tests       Enable/disable running already successful tests
  --test-arguments TEST_ARGS
                           Arguments passed in to the test suite program
  --coverage               Generate a code coverage report
  --no-run-tests           Disable running of tests. (Tests will still be
                           built.)
  --[no-]bench             Enable/disable benchmarking the package(s) in this
                           directory/configuration
  --benchmark-arguments BENCH_ARGS
                           Forward BENCH_ARGS to the benchmark suite. Supports
                           templates from `cabal bench`
  --no-run-benchmarks      Disable running of benchmarks. (Benchmarks will still
                           be built.)
  --exec CMD [ARGS]        Command and arguments to run after a successful build
  --only-configure         Only perform the configure step, not any builds.
                           Intended for tool usage, may break when used on
                           multiple packages at once!
  --reconfigure            Perform the configure step even if unnecessary.
                           Useful in some corner cases with custom Setup.hs
                           files
  --cabal-verbose          Ask Cabal to be verbose in its output
  --help                   Show this help text

Run 'stack --help' for global options that apply to all subcommands.
```
After that PR got merged (32b6da83595011dad9b8340dae86e65e33c273e5) the output is now
```
Usage: stack build [TARGET] [--dry-run] [--pedantic] [--fast]
                   [--ghc-options OPTION] [--flag PACKAGE:[-]FLAG]
                   ([--dependencies-only] | [--only-snapshot] |
                   [--only-dependencies]) ([--file-watch] | [--file-watch-poll])
                   [--exec CMD [ARGS]] [--only-configure] [--help]
  Build the package(s) in this directory/configuration

Available options:
  TARGET                   If none specified, use all packages
  --dry-run                Don't build anything, just prepare to
  --pedantic               Turn on -Wall and -Werror
  --fast                   Turn off optimizations (-O0)
  --ghc-options OPTION     Additional options passed to GHC
  --flag PACKAGE:[-]FLAG   Override flags set in stack.yaml (applies to local
                           packages and extra-deps)
  --dependencies-only      A synonym for --only-dependencies
  --only-snapshot          Only build packages for the snapshot database, not
                           the local database
  --only-dependencies      Only build packages that are dependencies of targets
                           on the command line
  --file-watch             Watch for changes in local files and automatically
                           rebuild. Ignores files in VCS boring/ignore file
  --file-watch-poll        Like --file-watch, but polling the filesystem instead
                           of using events
  --exec CMD [ARGS]        Command and arguments to run after a successful build
  --only-configure         Only perform the configure step, not any builds.
                           Intended for tool usage, may break when used on
                           multiple packages at once!
  --help                   Show this help text

Run 'stack --help' for global options that apply to all subcommands.
```
I haven’t gotten around to looking at the code.
Pinging @drwebb since it was his PR.

**Comment by drwebb** [comment]
Thanks for the report, the `hideMods` flag isn't being properly set in `configOptsParser` (I just replicated the lines for the docker and nix options parsers). I have tested a little fix, and it looks like I can do the same to the docker and nix commands as well.

**Comment by drwebb** [comment]
Okay, I've come up with another fix (the first way was printing the help too often) which should involve version of `addCommand` that takes in another parser to extract and append its help chunks.

**Comment by mgsloan** [comment]
It'd be good to have this fixed!  I can't test the `--help` output for the new `--split-objs` flag.

**Comment by drwebb** [comment]
Sorry, I dropped the ball on getting a simple fix pushed, but I just pushed a fix to master df9bc523b. In the end it was easier to not mess around with optparse-applicative, and instead just go for finer control over the flags that either show or hide the help text.

**Comment by mgsloan** [comment]
Cool, thanks!
